projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,154 @@
+}
+
+// reads the register and stores the value in the given array
+static StatusCode prv_read_register(I2CPort i2c_port, Ads1015Address i2c_addr, uint8_t reg,
+                                    uint8_t *read, size_t size_read) {
+  status_ok_or_return(i2c_write(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, &reg, sizeof(reg)));
+
+  status_ok_or_return(i2c_read(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, read, size_read));
+  return STATUS_CODE_OK;
+}
+
+// Switches to the the given channel by writing to config reg
+static StatusCode prv_set_channel(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_CONFIG,
+                                         CONFIG_REGISTER_MSB(channel), CONFIG_REGISTER_LSB));
+  storage->current_channel = channel;
+  storage->channel_enable[channel] = true;
+  return STATUS_CODE_OK;
+}
+
+// This function is registered as the callback for ALRT/RDY Pin.
+// reads and stores the conversion value in storage, and switches to the next enabled channel.
+static void prv_interrupt_handler(const GPIOAddress *address, void *context) {
+  Ads1015Storage *storage = context;
+
+  uint8_t read_conv_register[2];
+  prv_read_register(storage->i2c_port, storage->i2c_addr, ADS1015_ADDRESS_POINTER_CONV,
+                    read_conv_register, SIZEOF_ARRAY(read_conv_register));
+
+  uint8_t config_register[2];
+  prv_read_register(storage->i2c_port, storage->i2c_addr, ADS1015_ADDRESS_POINTER_CONFIG,
+                    config_register, SIZEOF_ARRAY(config_register));
+
+  Ads1015Channel current_channel = storage->current_channel;
+  // following line puts the two read bytes into an int16.
+  storage->channel_readings[current_channel] =
+      ((read_conv_register[0] << 8) | read_conv_register[1]);
+  // runs the users callback if not NULL
+  if (storage->channel_callback[current_channel] != NULL) {
+    storage->channel_callback[current_channel](storage->current_channel,
+                                               storage->callback_context[current_channel]);
+  }
+  // Update so that the ADS1015 reads from the next channel
+  do {
+    current_channel = (current_channel + 1) % NUM_ADS1015_CHANNELS;
+  } while (!storage->channel_enable[current_channel]);
+  prv_set_channel(storage, current_channel);
+}
+
+// initiates ads1015 by setting up registers and enabling ALRT/RDY Pin
+StatusCode ads1015_init(Ads1015Storage *storage, I2CPort i2c_port, Ads1015Address i2c_addr,
+                        GPIOAddress *ready_pin) {
+  if ((storage == NULL) || (ready_pin == NULL)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  storage->i2c_port = i2c_port;
+  storage->i2c_addr = i2c_addr;
+  storage->ready_pin = *ready_pin;
+  for (uint8_t channel = 0; channel < NUM_ADS1015_CHANNELS; channel++) {
```

Comment:
You could `memset` the whole storage to clear it.

Reply:
Will memsetting to zero make the pointers NULL and the bools false?

Reply:
I checked examples from the repo. It will do that.

Reply:
Just make sure you understand why it works in this case.